### PR TITLE
Feat/okx perpetuals v5 connector

### PR DESCRIPTION
**Pull request by tomasgaudino**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Implement OKX Perpetuals Connector according to #6775

**Tests performed by the developer**:
- All tests proposed in QA Section on https://hummingbot-foundation.notion.site/Perp-Connector-v2-1-57d8391eb54c40929f77067355fd551e

### CODE REVIEW

**Comment by nikspz** [comment]
image

Could you change example_pair to BTC-USDT? @tomasgaudino

**Comment by nikspz** [comment]
@tomasgaudino 
There's the issues I had on order_amount and position management now:

Steps:
1. set up simple perpetualMM strategy using okx_perpetual STRK-USDT and leverage 3
2. run for longtime
3. review logs

Actual:
order volume is wrong for CLOSE order
position not closed on exchange
got bot stuck on Creating stop loss order buy order to close short position
order amount do not matched order amount we set up (order_amount 7) 


Order amount bigger than order_amount set up

image
image


image


retest on MATIC-USDT Leverage 1 

Steps:
1. set up simple perpetualMM strategy using okx_perpetual MATIC-USDT and leverage 1
2. run for longtime
3. review logs

Actual:
wrong amount to CLOSE

image
image

**Comment by nikspz** [comment]
- Test 02414e6:
    - Cloned and installed branch
    - connect okx_perpetual API keys successfully
    - balance matched with exchange
    - Created strategy using okx perpetual
    - Started strategy, no issues showed
    - Reviewed orders placed successfully on exchange
    - Checked status successfully
    - Order cancellation - ok
    - Order filled events, position managment - ok
    - Data aggregate (matched with datadog)- ok
    - Test different token pairs - ok
    - Vertify fee calculation - ok
    - Data synced with exchage on ticker, status, order_book --live
        - order book amount not matched with exchange
        
        
[0227.zip](https://github.com/hummingbot/hummingbot/files/14420317/0227.zip)
[logs_conf_perpetual_market_making_1.log](https://github.com/hummingbot/hummingbot/files/14420303/logs_conf_perpetual_market_making_1.log)

image

**Comment by nikspz** [comment]
- commit 4694be4
- Test performed:
    - connected okx_perpetual succesfully
    - check balance ok
    - created/started perpetualMM without
    - order_book —live matched with exchange
image

For position managment, looks like that some position could be not correct (could be related to high volatility, `bot was able to recover`, placed stop loss according to config) 

[0228.zip](https://github.com/hummingbot/hummingbot/files/14439897/0228.zip)
[logs_conf_perpetual_market_making_1.log](https://github.com/hummingbot/hummingbot/files/14439876/logs_conf_perpetual_market_making_1.log)
[trades_conf_perpetual_market_making_1.csv](https://github.com/hummingbot/hummingbot/files/14439879/trades_conf_perpetual_market_making_1.csv)

```
2024-02-28 20:04:48,020 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Creating 1 LIMIT bid orders at (Size, Price): ['2E+1 MATIC, 0.9769 USDT'] to OPEN position.
2024-02-28 20:04:48,021 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Creating 1  LIMIT ask orders at (Size, Price): ['2E+1 MATIC, 0.9808 USDT'] to OPEN position.
2024-02-28 20:04:48,129 - 176165 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT BUY order 93027a12dac34fBCBMCUT9dc468dbdef for 20 to OPEN a MATIC-USDT position.
2024-02-28 20:04:48,144 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150688.0, "type": "OrderType.LIMIT", "trading_pair": "MATIC-USDT", "amount": "20", "price": "0.9769", "order_id": "93027a12dac34fBCBMCUT9dc468dbdef", "creation_timestamp": 1709150688.0, "exchange_order_id": "683162732866568221", "leverage": 5, "position": "OPEN", "event_name": "BuyOrderCreatedEvent", "event_source": "okx_perpetual"}
2024-02-28 20:04:48,307 - 176165 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT SELL order 93027a12dac34fBCSMCUT789c71dbe6b for 20 to OPEN a MATIC-USDT position.
2024-02-28 20:04:48,317 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150688.0, "type": "OrderType.LIMIT", "trading_pair": "MATIC-USDT", "amount": "20", "price": "0.9808", "order_id": "93027a12dac34fBCSMCUT789c71dbe6b", "creation_timestamp": 1709150688.0, "exchange_order_id": "683162733726400551", "leverage": 5, "position": "OPEN", "event_name": "SellOrderCreatedEvent", "event_source": "okx_perpetual"}
2024-02-28 20:06:23,220 - 176165 - hummingbot.connector.client_order_tracker - INFO - The BUY order 93027a12dac34fBCBMCUT9dc468dbdef amounting to 20/20 MATIC has been filled.
2024-02-28 20:06:23,221 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Maker buy order of 20 MATIC filled. [clock=2024-02-28 20:06:23+00:00]
2024-02-28 20:06:23,255 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150783.0, "order_id": "93027a12dac34fBCBMCUT9dc468dbdef", "trading_pair": "MATIC-USDT", "trade_type": "TradeType.BUY", "order_type": "OrderType.LIMIT", "price": "0.9769", "amount": "20", "trade_fee": {"percent": "0", "percent_token": "USDT", "flat_fees": [{"token": "USDT", "amount": "0.0039076"}]}, "exchange_trade_id": "131234854", "exchange_order_id": "683162732866568221", "leverage": 5, "position": "OPEN", "event_name": "OrderFilledEvent", "event_source": "okx_perpetual"}
2024-02-28 20:06:23,256 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Maker buy order 93027a12dac34fBCBMCUT9dc468dbdef (20 MATIC @ 0.9769 USDT) has been completely filled. [clock=2024-02-28 20:06:23+00:00]
2024-02-28 20:06:23,291 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150783.0, "order_id": "93027a12dac34fBCBMCUT9dc468dbdef", "base_asset": "MATIC", "quote_asset": "USDT", "base_asset_amount": "20", "quote_asset_amount": "19.5380", "order_type": "OrderType.LIMIT", "exchange_order_id": "683162732866568221", "event_name": "BuyOrderCompletedEvent", "event_source": "okx_perpetual"}
2024-02-28 20:06:23,291 - 176165 - hummingbot.connector.client_order_tracker - INFO - BUY order 93027a12dac34fBCBMCUT9dc468dbdef completely filled.
2024-02-28 20:06:24,001 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Canceling the limit order 93027a12dac34fBCSMCUT789c71dbe6b. [clock=2024-02-28 20:06:24+00:00]
2024-02-28 20:06:24,001 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - Initiated cancelation of sell order 93027a12dac34fBCSMCUT789c71dbe6b in favour of take profit order.
2024-02-28 20:06:24,441 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150784.0, "order_id": "93027a12dac34fBCSMCUT789c71dbe6b", "exchange_order_id": "683162733726400551", "event_name": "OrderCancelledEvent", "event_source": "okx_perpetual"}
2024-02-28 20:06:24,442 - 176165 - hummingbot.connector.client_order_tracker - INFO - Successfully canceled order 93027a12dac34fBCSMCUT789c71dbe6b.
2024-02-28 20:06:30,001 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Creating 1  LIMIT ask orders at (Size, Price): ['2E+1 MATIC, 0.9788 USDT'] to CLOSE position.
2024-02-28 20:06:30,086 - 176165 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT SELL order 93027a12dac34fBCSMCUT0c7742e27af for 20 to CLOSE a MATIC-USDT position.
2024-02-28 20:06:30,095 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150790.0, "type": "OrderType.LIMIT", "trading_pair": "MATIC-USDT", "amount": "20", "price": "0.9788", "order_id": "93027a12dac34fBCSMCUT0c7742e27af", "creation_timestamp": 1709150790.0, "exchange_order_id": "683163160593301520", "leverage": 5, "position": "CLOSE", "event_name": "SellOrderCreatedEvent", "event_source": "okx_perpetual"}
2024-02-28 20:07:30,003 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Canceling the limit order 93027a12dac34fBCSMCUT0c7742e27af. [clock=2024-02-28 20:07:30+00:00]
2024-02-28 20:07:30,004 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - Canceling the limit order 93027a12dac34fBCSMCUT0c7742e27af to renew stop loss.
2024-02-28 20:07:30,154 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150850.0, "order_id": "93027a12dac34fBCSMCUT0c7742e27af", "exchange_order_id": "683163160593301520", "event_name": "OrderCancelledEvent", "event_source": "okx_perpetual"}
2024-02-28 20:07:30,155 - 176165 - hummingbot.connector.client_order_tracker - INFO - Successfully canceled order 93027a12dac34fBCSMCUT0c7742e27af.
2024-02-28 20:08:01,000 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Creating 1 LIMIT bid orders at (Size, Price): ['2E+1 MATIC, 0.9749 USDT'] to CLOSE position.
2024-02-28 20:08:01,091 - 176165 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT BUY order 93027a12dac34fBCBMCUTb78504f093e for 20 to CLOSE a MATIC-USDT position.
2024-02-28 20:08:01,103 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150881.0, "type": "OrderType.LIMIT", "trading_pair": "MATIC-USDT", "amount": "20", "price": "0.9749", "order_id": "93027a12dac34fBCBMCUTb78504f093e", "creation_timestamp": 1709150881.0, "exchange_order_id": "683163542283354126", "leverage": 5, "position": "CLOSE", "event_name": "BuyOrderCreatedEvent", "event_source": "okx_perpetual"}
2024-02-28 20:08:01,319 - 176165 - hummingbot.connector.client_order_tracker - INFO - The BUY order 93027a12dac34fBCBMCUTb78504f093e amounting to 20/20 MATIC has been filled.
2024-02-28 20:08:01,319 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Maker buy order of 20 MATIC filled. [clock=2024-02-28 20:08:01+00:00]
2024-02-28 20:08:01,349 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150881.0, "order_id": "93027a12dac34fBCBMCUTb78504f093e", "trading_pair": "MATIC-USDT", "trade_type": "TradeType.BUY", "order_type": "OrderType.LIMIT", "price": "0.9749", "amount": "20", "trade_fee": {"percent": "0", "percent_token": "USDT", "flat_fees": [{"token": "USDT", "amount": "0.0038996"}]}, "exchange_trade_id": "131235614", "exchange_order_id": "683163542283354126", "leverage": 5, "position": "CLOSE", "event_name": "OrderFilledEvent", "event_source": "okx_perpetual"}
2024-02-28 20:08:01,350 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Maker buy order 93027a12dac34fBCBMCUTb78504f093e (20 MATIC @ 0.9749 USDT) has been completely filled. [clock=2024-02-28 20:08:01+00:00]
2024-02-28 20:08:01,416 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150881.0, "order_id": "93027a12dac34fBCBMCUTb78504f093e", "base_asset": "MATIC", "quote_asset": "USDT", "base_asset_amount": "20", "quote_asset_amount": "19.4980", "order_type": "OrderType.LIMIT", "exchange_order_id": "683163542283354126", "event_name": "BuyOrderCompletedEvent", "event_source": "okx_perpetual"}
2024-02-28 20:08:01,416 - 176165 - hummingbot.connector.client_order_tracker - INFO - BUY order 93027a12dac34fBCBMCUTb78504f093e completely filled.
2024-02-28 20:08:18,002 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - Creating stop loss sell order to close long position.
2024-02-28 20:08:18,002 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Creating 1  LIMIT ask orders at (Size, Price): ['4E+1 MATIC, 0.9685 USDT'] to CLOSE position.
2024-02-28 20:08:18,107 - 176165 - hummingbot.connector.client_order_tracker - INFO - The SELL order 93027a12dac34fBCSMCUT8a95b78dd39 amounting to 40/40 MATIC has been filled.
2024-02-28 20:08:18,107 - 176165 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (MATIC-USDT) Maker sell order of 40 MATIC filled. [clock=2024-02-28 20:08:18+00:00]
2024-02-28 20:08:18,125 - 176165 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1709150898.0, "order_id": "93027a12dac34fBCSMCUT8a95b78dd39", "trading_pair": "MATIC-USDT", "trade_type": "TradeType.SELL", "order_type": "OrderType.LIMIT", "price": "0.9732", "amount": "40", "trade_fee": {"percent": "0", "percent_token": "USDT", "flat_fees": [{"token": "USDT", "amount": "0.019464"}]}, "exchange_trade_id": "131235756", "exchange_order_id": "683163613603299328", "leverage": 5, "position": "CLOSE", "event_name": "OrderFilledEvent", "event_source": "okx_perpetual"}
2024-02-28 20:08:18,126 - 176165 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT SELL order 93027a12dac34fBCSMCUT8a95b78dd39 for 40 to CLOSE a MATIC-USDT position.
```

image
image


    - pending: longrun test

**Comment by nikspz** [comment]
trade PNL inconsistent leads to killswitch triggered

Steps:
- connect okx_perpetual API keys
- check balance
- Create / Start perpetualMM strategy using okx perpetual
- wait till 10-20 trades filled
- check balance again
- Review Trade PNL is inconsistent which could lead to killswitch triggered


[logs_conf_perpetual_market_making_4.log](https://github.com/hummingbot/hummingbot/files/14448978/logs_conf_perpetual_market_making_4.log)
[0228-1.zip](https://github.com/hummingbot/hummingbot/files/14448976/0228-1.zip)


image

**Comment by tomasgaudino** [comment]
It seems that last two fixes corrected both issues. I couldn't get a history command screenshot because I had to reboot aws ec2 instance, but here there are the results:

[2024-03-03.zip](https://github.com/hummingbot/hummingbot/files/14482097/2024-03-03.zip)

**Comment by nikspz** [comment]
Test performed:
- Cloned and installed feature branch
- Started the client
- Connected the connector API keys = OK
- Balance output = OK
- Create/Import/Start a strategy = OK
- Check status and status --live match with exchange = OK
- Order cancellation = OK
- Ticket and Order book --live match with exchange = OK
- Order filled events = OK
- Vertify fee calculation = OK
- Data collected by datadog = OK
- Test different token pairs = OK
- Position retrieval = OK

Pending: longrun test on v2 strategies

**Review by nikspz — APPROVED** [review]
Test performed:
- Cloned and installed feature branch
- Started the client Successfully
- Manually built docker image succesfully
- Connected the connector API keys = OK
- Balance output = OK
- Create/Import/Start a strategy = OK
- Check status and status --live match with exchange = OK
- Order cancellation = OK
- Ticket and Order book --live match with exchange = OK
- Order filled events = OK
- Vertify fees matched with exchange
- Test different token pairs = OK
- Position retrieval = OK
- longrun test on d-manv3 using okx_perpetual candles = OK

Note:
Last Data for okx_perpetual is not collected by datadog.

File: hummingbot/connector/derivative/okx_perpetual/okx_perpetual_constants.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+# -------------------------------------------
+# WEB UTILS ENDPOINTS
+# The structure is REST_url = {method: GET/POST, endpoint: /api/v5/...} since for the same endpoint you can have
+# different methods. This is also useful for rate limit ids.
+# -------------------------------------------
+# REST API Public Endpoints
+REST_LATEST_SYMBOL_INFORMATION = {METHOD: GET,
+                                  ENDPOINT: f"/api/{REST_API_VERSION}/market/tickers"}
+REST_ORDER_BOOK = {METHOD: GET,
```
Comment (cardosofede) [inline hummingbot/connector/derivative/okx_perpetual/okx_perpetual_constants.py:98]: in the next version simplify the constants to only hold the url

File: hummingbot/connector/derivative/okx_perpetual/okx_perpetual_web_utils.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    rate_limits = []
+    domain = CONSTANTS.DEFAULT_DOMAIN
+    rate_limits.extend(_build_websocket_rate_limits(domain))
+    rate_limits.extend(_build_public_rate_limits())
+    rate_limits.extend(_build_private_rate_limits(trading_pairs))
+
+    return rate_limits
+
+
+def _build_websocket_rate_limits(domain: str) -> List[RateLimit]:
```
Comment (cardosofede) [inline hummingbot/connector/derivative/okx_perpetual/okx_perpetual_web_utils.py:141]: in the next version simplify the construction in one list, as is done in OKX spot. Also move it to constants.

File: hummingbot/connector/derivative/okx_perpetual/okx_perpetual_auth.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        headers = {}
+        if request.headers is not None:
+            headers.update(request.headers)
+        headers.update(self.authentication_headers(request=request))
+        request.headers = headers
+
+        return request
+
+    @staticmethod
+    def get_path_from_url(url: str) -> str:
```
Comment (cardosofede) [inline hummingbot/connector/derivative/okx_perpetual/okx_perpetual_auth.py:87]: this method is not used

File: hummingbot/connector/derivative/okx_perpetual/okx_perpetual_user_stream_data_source.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    @property
+    def last_recv_time(self) -> float:
+        """
+        Returns the time of the last received message
+
+        :return: the timestamp of the last received message in seconds
+        """
+        t = 0.0
+        if len(self._ws_assistants) > 0:
+            t = min([wsa.last_recv_time for wsa in self._ws_assistants])
```
Comment (cardosofede) [inline hummingbot/connector/derivative/okx_perpetual/okx_perpetual_user_stream_data_source.py:40]: shouldn't be max()?

File: hummingbot/connector/derivative/okx_perpetual/okx_perpetual_derivative.py [hunk 5]
```diff
@@ -0,0 +1,886 @@
...
+
+    def supported_position_modes(self) -> List[PositionMode]:
+        return [PositionMode.ONEWAY, PositionMode.HEDGE]
+
+    def _is_request_exception_related_to_time_synchronizer(self, request_exception: Exception):
+        error_description = str(request_exception)
+        is_time_synchronizer_related = '"code":"50113"' in error_description
+        return is_time_synchronizer_related
+
+    def _is_order_not_found_during_status_update_error(self, status_update_exception: Exception) -> bool:
```
Comment (cardosofede) [inline hummingbot/connector/derivative/okx_perpetual/okx_perpetual_derivative.py:147]: implement this in the next version

**Review by cardosofede — APPROVED** [review]
LGTM!

**Comment by tomasgaudino** [comment]
hello @david-hummingbot ! sure, today i'll be updating it

**Pull request merged**